polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFilter.java

```diff
@@ -0,0 +1,92 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.ratelimiter;
+
+import jakarta.annotation.Priority;
+import jakarta.servlet.Filter;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.ServletRequest;
+import jakarta.servlet.ServletResponse;
+import jakarta.servlet.http.HttpServletResponse;
+import jakarta.ws.rs.Priorities;
+import jakarta.ws.rs.core.Response;
+import java.io.IOException;
+import java.time.Clock;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.polaris.core.context.CallContext;
+import org.apache.polaris.core.context.RealmContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Request filter that returns a 429 Too Many Requests if the rate limiter says so */
+@Priority(Priorities.AUTHORIZATION + 1)
+public class RateLimiterFilter implements Filter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RateLimiterFilter.class);
+  private static final RateLimiter NO_OP_LIMITER = new NoOpRateLimiter();
+  private static final RateLimiter ALWAYS_REJECT_LIMITER =
+      new TokenBucketRateLimiter(0, 0, Clock.systemUTC());
+
+  private final RateLimiterConfig config;
+  private final Map<String, Future<RateLimiter>> perRealmLimiters = new ConcurrentHashMap<>();
+
+  public RateLimiterFilter(RateLimiterConfig config) {
+    this.config = config;
+  }
+
+  /** Returns a 429 if the rate limiter says so. Otherwise, forwards the request along. */
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+      throws IOException, ServletException {
+    RealmContext realmContext = CallContext.getCurrentContext().getRealmContext();
+    RateLimiter rateLimiter = maybeBlockToGetRateLimiter(realmContext);
+    if (!rateLimiter.tryAcquire()) {
+      ((HttpServletResponse) response).setStatus(Response.Status.TOO_MANY_REQUESTS.getStatusCode());
+      return;
+    }
+    chain.doFilter(request, response);
+  }
+
+  private RateLimiter maybeBlockToGetRateLimiter(RealmContext realmContext) {
+    try {
+      return perRealmLimiters
```

Comment:
Are we going to introduce finer granularity limiters later? For example, rate limiters based on principal/Ip address?

Reply:
+1 to this, maybe we can introduce a thin type to cover our bases here. 

So...
```
public record RateLimiterInfo(RealmContext realmContext) {}

. . .

private RateLimiter maybeBlockToGetRateLimiter(RateLimiterInfo rateLimiterInfo) { ... }
private final Map<RateLimiterInfo, Future<RateLimiter>> rateLimiters = new ConcurrentHashMap<>();
// etc.

...
```

This might let us add more fine grained rate limiters later on without having to hack up the APIs.

Reply:
Added a layer above the key called `RateLimiterKey`. To answer the original question, it should be straight forward to rate limit on different properties and having a custom key class makes it so others don't have to refactor their code when we do so.